File: packages/python/src/brainpick/core/bundle.py
"""Bundle scanning: documents, tolerant metadata, resolved links (spec/20)."""
from __future__ import annotations

import posixpath
import re
from dataclasses import dataclass, field
from datetime import date, datetime, timezone
from fnmatch import fnmatch
from pathlib import Path

from brainpick.core.canonical import sha256_hex
from brainpick.core.frontmatter import split_frontmatter
from brainpick.core.links import RawLink, extract_links

ALWAYS_EXCLUDED_DIRS = {".brainpick", ".git", "_temp", "node_modules"}
RESERVED_NAMES = {"index.md", "log.md", "skilltree.md"}
SKILL_TYPES = {"skill", "playbook"}  # spec/85: a skill by type, wherever it lives
_H1 = re.compile(r"^# +(.+?)\s*$", re.MULTILINE)


@dataclass(frozen=True)
class ResolvedLink:
    kind: str
    target: str
    text: str


@dataclass(frozen=True)
class Ghost:
    target: str


@dataclass
class Document:
    path: str
    sha256: str
    size: int
    title: str
    type: str | None
    about: str | None
    description: str | None
    tags: list[str]
    timestamp: str | None
    reserved: bool
    body: str
    links: list[ResolvedLink] = field(default_factory=list)
    ghosts: list[Ghost] = field(default_factory=list)
    skill: bool = False                                   # spec/20 Skills and frontmatter edges
    depends_on: list[str] = field(default_factory=list)   # RESOLVED prerequisites, declared order
    tools: list[str] = field(default_factory=list)        # declared tool paths, as written


def is_skill(type_value) -> bool:
    """A doc is a skill by its `type` — `skill` or the older `playbook`, trimmed,
    case-insensitive — never by folder (spec/85)."""
    return type_value is not None and str(type_value).strip().lower() in SKILL_TYPES


def _normalize_list(value) -> list[str]:
    """depends_on / tools: absent -> [], scalar wraps, values coerce to strings."""
    if value is None:
        return []
    if isinstance(value, list):
        return [str(v) for v in value if v is not None]
    return [str(value)]


def _normalize_timestamp(value) -> str | None:
    if value is None:
        return None
    if isinstance(value, datetime):
        if value.tzinfo is None:
            value = value.replace(tzinfo=timezone.utc)
        return value.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    if isinstance(value, date):
        return value.strftime("%Y-%m-%d")
    return str(value)


def _normalize_tags(value) -> list[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return [str(v) for v in value]
    return [str(value)]


def resolve_doc_target(source: str, target: str, file_set: set[str]) -> str | None:
    """A frontmatter doc target (depends_on): rooted from the bundle root first —
    `a/b`, `a/b.md`, `a/b/index.md` — then relative to the source's directory,
    as given, then with `.md` appended (spec/20)."""
    target = str(target or "").strip()
    if not target:
        return None
    base = target.lstrip("/")
    for cand in (base, base + ".md", posixpath.join(base, "index.md")):
        if cand in file_set:
            return cand
    if target.startswith("/"):
        return None
    joined = posixpath.normpath(posixpath.join(posixpath.dirname(source), target))
    if joined.startswith(".."):
        return None
    for cand in (joined, joined + ".md"):
        if cand in file_set:
            return cand
    return None


def title_of(meta: dict, body: str, path: str) -> str:
    if meta.get("title") is not None:
        return str(meta["title"])
    m = _H1.search(body)
    if m:
        return m.group(1)
    stem = posixpath.basename(path).rsplit(".", 1)[0]
    return stem.replace("-", " ").replace("_", " ")


def _collect_files(root: Path, include: tuple[str, ...], exclude: tuple[str, ...]) -> list[str]:
    files: set[str] = set()
    for pattern in include:
        for p in root.glob(pattern):
            if not p.is_file():
                continue
            rel = p.relative_to(root).as_posix()
            parts = rel.split("/")
            if any(part in ALWAYS_EXCLUDED_DIRS for part in parts[:-1]):
                continue
            if any(fnmatch(rel, ex) for ex in exclude):
                continue
            files.add(rel)
    return sorted(files)


def build_stem_maps(paths: list[str]) -> tuple[dict[str, list[str]], dict[str, list[str]]]:
    """(stems, case-insensitive stems) for wikilink resolution — shared by scan
    and the algorithmic T3 derivation, so both resolve identically (spec/40)."""
    stems: dict[str, list[str]] = {}
    stems_ci: dict[str, list[str]] = {}
    for p in paths:
        stem = posixpath.basename(p).rsplit(".", 1)[0]
        stems.setdefault(stem, []).append(p)
        stems_ci.setdefault(stem.lower(), []).append(p)
    return stems, stems_ci


def resolve_link(source: str, raw: RawLink, file_set: set[str],
                 stems: dict[str, list[str]], stems_ci: dict[str, list[str]]) -> str | None:
    if raw.kind == "wikilink":
        exact = stems.get(raw.target, [])
        if len(exact) == 1:
            return exact[0]
        ci = stems_ci.get(raw.target.lower(), [])
        if len(ci) == 1:
            return ci[0]
        return None

    target = raw.target
    if target.startswith("/"):
        base = target.lstrip("/")
        for cand in (base, base + ".md", posixpath.join(base, "index.md")):
            if cand in file_set:
                return cand
        return None

    joined = posixpath.normpath(posixpath.join(posixpath.dirname(source), target))
    if joined.startswith(".."):
        return None
    for cand in (joined, joined + ".md"):
        if cand in file_set:
            return cand
    return None


def scan(root: str | Path, include: tuple[str, ...] = ("**/*.md",),
         exclude: tuple[str, ...] = ()) -> list[Document]:
    root = Path(root)
    paths = _collect_files(root, include, exclude)
    file_set = set(paths)
    stems, stems_ci = build_stem_maps(paths)

    docs: list[Document] = []
    for path in paths:
        raw_bytes = (root / path).read_bytes()
        meta, body = split_frontmatter(raw_bytes.decode("utf-8", errors="replace"))

        links: list[ResolvedLink] = []
        ghosts: list[Ghost] = []
        for raw in extract_links(body):
            resolved = resolve_link(path, raw, file_set, stems, stems_ci)
            if resolved == path:
                continue  # self-links are dropped
            if resolved is None:
                ghosts.append(Ghost(target=raw.target))
            else:
                links.append(ResolvedLink(kind=raw.kind, target=resolved, text=raw.text))

        reserved = posixpath.basename(path) in RESERVED_NAMES
        skill = is_skill(meta.get("type")) and not reserved
        depends_on: list[str] = []
        tools: list[str] = []
        if skill:
            for declared in _normalize_list(meta.get("depends_on")):
                resolved = resolve_doc_target(path, declared, file_set)
                if resolved == path:
                    continue  # self-dependencies are dropped, like self-links
                if resolved is None:
                    ghosts.append(Ghost(target=declared))
                elif resolved not in depends_on:
                    depends_on.append(resolved)
            tools = _normalize_list(meta.get("tools"))

        docs.append(Document(
            path=path,
            sha256=sha256_hex(raw_bytes),
            size=len(raw_bytes),
            title=title_of(meta, body, path),
            type=None if meta.get("type") is None else str(meta["type"]),
            about=None if meta.get("about") is None else str(meta["about"]),
            description=None if meta.get("description") is None else str(meta["description"]),
            tags=_normalize_tags(meta.get("tags")),
            timestamp=_normalize_timestamp(meta.get("timestamp")),
            reserved=reserved,
            body=body,
            links=links,
            ghosts=ghosts,
            skill=skill,
            depends_on=depends_on,
            tools=tools,
        ))
    return docs

File: packages/node/src/core/bundle.ts
/** Bundle scanning: documents, tolerant metadata, resolved links (spec/20). */
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";

import { cmpStr, sha256Hex } from "./canonical";
import { splitFrontmatter } from "./frontmatter";
import { extractLinks, type RawLink } from "./links";
import { PY_SPACE_CLASS } from "./pyfmt";
import { pyStr, YamlTimestamp } from "./yaml11";

export const ALWAYS_EXCLUDED_DIRS = new Set([".brainpick", ".git", "_temp", "node_modules"]);
export const RESERVED_NAMES = new Set(["index.md", "log.md"]);

export const DEFAULT_INCLUDE: readonly string[] = ["**/*.md"];

// Python: ^# +(.+?)\s*$ MULTILINE — `.` is [^\n], `\s` the Python class.
const H1 = new RegExp(`^# +([^\\n]+?)[${PY_SPACE_CLASS}]*$`, "mu");

export interface ResolvedLink {
  kind: "link" | "wikilink";
  target: string;
  text: string;
}

export interface Ghost {
  target: string;
}

export interface Document {
  path: string;
  sha256: string;
  size: number;
  title: string;
  type: string | null;
  description: string | null;
  tags: string[];
  timestamp: string | null;
  reserved: boolean;
  body: string;
  links: ResolvedLink[];
  ghosts: Ghost[];
}

export function normalizeTimestamp(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  if (value instanceof YamlTimestamp) return value.normalized();
  return pyStr(value);
}

export function normalizeTags(value: unknown): string[] {
  if (value === null || value === undefined) return [];
  if (Array.isArray(value)) return value.map(pyStr);
  return [pyStr(value)];
}

function titleOf(meta: Record<string, unknown>, body: string, path: string): string {
  if (meta["title"] !== null && meta["title"] !== undefined) return pyStr(meta["title"]);
  const m = H1.exec(body);
  if (m) return m[1]!;
  const stem = fileStem(posixBasename(path));
  return stem.replace(/-/g, " ").replace(/_/g, " ");
}

// ---------------------------------------------------------------------------
// posixpath ports (bundle-relative paths only ever use POSIX separators).

export function posixBasename(p: string): string {
  const i = p.lastIndexOf("/");
  return i === -1 ? p : p.slice(i + 1);
}

export function posixDirname(p: string): string {
  const i = p.lastIndexOf("/");
  if (i === -1) return "";
  let head = p.slice(0, i + 1);
  if (head !== "" && !/^\/+$/.test(head)) head = head.replace(/\/+$/, "");
  return head;
}

export function posixJoin(a: string, b: string): string {
  if (b.startsWith("/")) return b;
  if (a === "" || a.endsWith("/")) return a + b;
  return a + "/" + b;
}

/** posixpath.normpath, ported faithfully — including the '..'-escape shapes
 * that _resolve turns into ghosts. */
export function posixNormpath(path: string): string {
  if (path === "") return ".";
  const initialSlashes = path.startsWith("/")
    ? path.startsWith("//") && !path.startsWith("///")
      ? 2
      : 1
    : 0;
  const comps: string[] = [];
  for (const comp of path.split("/")) {
    if (comp === "" || comp === ".") continue;
    if (
      comp !== ".." ||
      (initialSlashes === 0 && comps.length === 0) ||
      (comps.length > 0 && comps[comps.length - 1] === "..")
    ) {
      comps.push(comp);
    } else if (comps.length > 0) {
      comps.pop();
    }
  }
  const joined = "/".repeat(initialSlashes) + comps.join("/");
  return joined || ".";
}

/** Python `posixpath.basename(p).rsplit(".", 1)[0]`. */
function fileStem(name: string): string {
  const i = name.lastIndexOf(".");
  return i === -1 ? name : name.slice(0, i);
}

// ---------------------------------------------------------------------------
// File collection. The 0.1 include set is fixed (`**/*.md`), so the tree is
// hand-walked instead of pulling in a glob dependency. Mirrors
// pathlib.Path.glob on Python 3.10–3.12: dotfiles are matched, directory
// symlinks are not followed, file symlinks count via is_file().

function collectFiles(root: string, include: readonly string[], exclude: readonly string[]): string[] {
  for (const pattern of include) {
    if (pattern !== "**/*.md") {
      throw new Error(`unsupported include pattern ${JSON.stringify(pattern)} — 0.1 fixes the include set to **/*.md`);
    }
  }
  if (include.length === 0) return [];
  const files: string[] = [];
  walk(root, "", files);
  const excludeRes = exclude.map(fnmatchToRegExp);
  return files.filter((rel) => !excludeRes.some((re) => re.test(rel))).sort(cmpStr);
}

function walk(dir: string, rel: string, out: string[]): void {
  let entries;
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const entry of entries) {
    const childRel = rel === "" ? entry.name : `${rel}/${entry.name}`;
    let isFile = entry.isFile();
    let isDir = entry.isDirectory();
    if (entry.isSymbolicLink()) {
      isDir = false; // Path.glob("**") does not follow directory symlinks
      try {
        isFile = statSync(join(dir, entry.name)).isFile();
      } catch {
        isFile = false;
      }
    }
    if (isDir) {
      if (!ALWAYS_EXCLUDED_DIRS.has(entry.name)) walk(join(dir, entry.name), childRel, out);
    } else if (isFile && entry.name.endsWith(".md")) {
      out.push(childRel);
    }
  }
}

/** fnmatch.fnmatch, minimally: `*` (crosses `/`), `?`, `[seq]`. The exclude list
 * reaches it here; the timeline's include/exclude matching (spec/90) reuses it. */
export function fnmatchToRegExp(pattern: string): RegExp {
  let re = "";
  for (let i = 0; i < pattern.length; i++) {
    const ch = pattern[i]!;
    if (ch === "*") re += ".*";
    else if (ch === "?") re += ".";
    else if (ch === "[") {
      const close = pattern.indexOf("]", i + 2);
      if (close === -1) {
        re += "\\[";
      } else {
        let seq = pattern.slice(i + 1, close);
        if (seq.startsWith("!")) seq = "^" + seq.slice(1);
        re += "[" + seq.replace(/\\/g, "\\\\") + "]";
        i = close;
      }
    } else re += ch.replace(/[.+^${}()|\\/[\]]/g, "\\$&");
  }
  return new RegExp(`^(?:${re})$`, "su");
}

// ---------------------------------------------------------------------------

function resolve(
  source: string,
  raw: RawLink,
  fileSet: Set<string>,
  stems: Map<string, string[]>,
  stemsCi: Map<string, string[]>,
): string | null {
  if (raw.kind === "wikilink") {
    const exact = stems.get(raw.target) ?? [];
    if (exact.length === 1) return exact[0]!;
    const ci = stemsCi.get(raw.target.toLowerCase()) ?? [];
    if (ci.length === 1) return ci[0]!;
    return null;
  }

  const target = raw.target;
  if (target.startsWith("/")) {
    const base = target.replace(/^\/+/, "");
    for (const cand of [base, base + ".md", posixJoin(base, "index.md")]) {
      if (fileSet.has(cand)) return cand;
    }
    return null;
  }

  const joined = posixNormpath(posixJoin(posixDirname(source), target));
  if (joined.startsWith("..")) return null;
  for (const cand of [joined, joined + ".md"]) {
    if (fileSet.has(cand)) return cand;
  }
  return null;
}

export function scan(
  root: string,
  include: readonly string[] = DEFAULT_INCLUDE,
  exclude: readonly string[] = [],
): Document[] {
  const paths = collectFiles(root, include, exclude);
  const fileSet = new Set(paths);

  const stems = new Map<string, string[]>();
  const stemsCi = new Map<string, string[]>();
  for (const p of paths) {
    const stem = fileStem(posixBasename(p));
    let list = stems.get(stem);
    if (!list) stems.set(stem, (list = []));
    list.push(p);
    const lower = stem.toLowerCase();
    let ciList = stemsCi.get(lower);
    if (!ciList) stemsCi.set(lower, (ciList = []));
    ciList.push(p);
  }

  const docs: Document[] = [];
  for (const path of paths) {
    const rawBytes = readFileSync(join(root, path));
    // Buffer.toString("utf8") replaces malformed sequences with U+FFFD, like
    // Python's errors="replace".
    const [meta, body] = splitFrontmatter(rawBytes.toString("utf8"));

    const links: ResolvedLink[] = [];
    const ghosts: Ghost[] = [];
    for (const raw of extractLinks(body)) {
      const resolved = resolve(path, raw, fileSet, stems, stemsCi);
      if (resolved === path) continue; // self-links are dropped
      if (resolved === null) ghosts.push({ target: raw.target });
      else links.push({ kind: raw.kind, target: resolved, text: raw.text });
    }

    docs.push({
      path,
      sha256: sha256Hex(rawBytes),
      size: rawBytes.length,
      title: titleOf(meta, body, path),
      type: meta["type"] === null || meta["type"] === undefined ? null : pyStr(meta["type"]),
      description:
        meta["description"] === null || meta["description"] === undefined ? null : pyStr(meta["description"]),
      tags: normalizeTags(meta["tags"]),
      timestamp: normalizeTimestamp(meta["timestamp"]),
      reserved: RESERVED_NAMES.has(posixBasename(path)),
      body,
      links,
      ghosts,
    });
  }
  return docs;
}
